craft_providers/multipass/multipass_instance.py

```diff
@@ -0,0 +1,387 @@
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Multipass Instance."""
+import io
+import logging
+import os
+import pathlib
+import subprocess
+import sys
+from typing import Any, Dict, List, Optional
+
+from .. import Executor
+from .errors import MultipassError
+from .multipass import Multipass
+
+logger = logging.getLogger(__name__)
+
+
+def _get_host_gid():
+    if sys.platform == "win32":
+        return 0
+
+    return os.getgid()
+
+
+def _get_host_uid():
+    if sys.platform == "win32":
+        return 0
+
+    return os.getuid()
+
+
+def _finalize_command(
+    command: List[str],
+    *,
+    env: Optional[Dict[str, str]] = None,
+) -> List[str]:
+    """Finalize command to run.
+
+    Account for environment (if any) and Multipass requirements to run as
+    root.
+
+    :param command: Command to execute.
+    :param env: Environment flags to set.
+
+    :returns: List of command strings.
+    """
+    final_cmd = ["sudo", "-H", "--"]
+
+    if env is not None:
+        env_args = [f"{k}={v}" for k, v in env.items()]
+        final_cmd += ["env", *env_args]
+
+    final_cmd += command
+
+    return final_cmd
+
+
+class MultipassInstance(Executor):
+    """Multipass Instance Lifecycle.
+
+    :param name: Name of multipass instance.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        multipass: Optional[Multipass] = None,
+    ):
+        super().__init__()
+
+        self.name = name
+
+        if multipass is not None:
+            self._multipass = multipass
+        else:
+            self._multipass = Multipass()
+
+    def create_file(
+        self,
+        *,
+        destination: pathlib.Path,
+        content: bytes,
+        file_mode: str,
+        group: str = "root",
+        user: str = "root",
+    ) -> None:
+        """Create file with content and file mode.
+
+        Multipass transfers data as "ubuntu" user, forcing us to first copy a
+        file to a temporary location before moving to a (possibly) root-owned
+        location and with appropriate permissions.
+
+        :param destination: Path to file.
+        :param content: Contents of file.
+        :param file_mode: File mode string (e.g. '0644').
+        :param group: File group owner/id.
+        :param user: File user owner/id.
+        """
+        stream = io.BytesIO(content)
```

Comment:
can BytesIO be the required input for this method?

Reply:
Sounds like an excellent idea.

Reply:
Will fix Wednesday

Reply:
done